test/project_types/extension/extension_project_test.rb

```diff
@@ -61,6 +61,11 @@ def test_write_env_file_persists_resource_url
       assert_equal("/cart/13:1", ExtensionProject.current.resource_url)
     end
 
+    def test_write_env_file_persists_shop
+      ExtensionProject.write_env_file(**valid_env_file_attributes_with(shop: "johndoe.myshopify.com"))
+      assert_equal("johndoe.myshopify.com", ExtensionProject.current.env.shop)
```

Comment:
Q: Are we sure that `ExtensionProject.current.env` loads the data from the `.env` file?
In other words, is it possible for the data to come from some in-memory cache, and be different than the data in the `.env` file?

Reply:
`#env` lazy loads the `.env` file, which is part of the pain. Combine that with the fact that we are loading this from disk in tests as well from a directory that is reused by several test and you can easily get yourself in trouble. It's absolutely possible that data in-memory and on disk differ. I lost hours on figuring out how to structure the code to ensure that this is not a problem for our particular use case. That said, I fully expect this to be a recurring problem. Environment handling has to be fundamentally rethought and implemented much better. Starting with not always overriding the entire env file but applying deltas. This would dramatically lower the chance of on-disk and in-memory representation differing.